pkg/openstack/routes.go

```diff
@@ -35,97 +36,127 @@ import (
 	if err != nil {
 		return nil, err
 	}
 
 	mc := metrics.NewMetricContext("router", "get")
-	router, err := routers.Get(r.network, r.opts.RouterID).Extract()
+	router, err := routers.Get(r.network, r.os.routeOpts.RouterID).Extract()
 	if mc.ObserveRequest(err) != nil {
 		return nil, err
 	}
 
 	var routes []*cloudprovider.Route
 	for _, item := range router.Routes {
-		nodeName, foundNode := nodeNamesByAddr[item.NextHop]
-		if !foundNode {
-			nodeName = types.NodeName(item.NextHop)
-		}
+		nodeName, foundNode := getNodeNameByAddr(item.NextHop, nodes)
 		route := cloudprovider.Route{
 			Name:            item.DestinationCIDR,
-			TargetNode:      nodeName, //contains the nexthop address if node was not found
+			TargetNode:      nodeName, //contains the nexthop address if node name was not found
 			Blackhole:       !foundNode,
 			DestinationCIDR: item.DestinationCIDR,
 		}
 		routes = append(routes, &route)
 	}
 
+	// detect router's private network ID for further VM ports filtering
+	r.Lock()
+	r.networkIDs, err = getRouterNetworkIDs(r.network, r.os.routeOpts.RouterID)
+	r.Unlock()
+	if err != nil {
+		return nil, err
+	}
+
 	return routes, nil
 }
 
-func foreachServer(client *gophercloud.ServiceClient, opts servers.ListOptsBuilder, handler func(*servers.Server) (bool, error)) error {
-	mc := metrics.NewMetricContext("server", "list")
-	pager := servers.List(client, opts)
+func getRouterNetworkIDs(network *gophercloud.ServiceClient, routerID string) ([]string, error) {
+	opts := ports.ListOpts{
+		DeviceID: routerID,
+	}
+	mc := metrics.NewMetricContext("port", "list")
+	pages, err := ports.List(network, opts).AllPages()
+	if mc.ObserveRequest(err) != nil {
+		return nil, err
+	}
+	ports, err := ports.ExtractPorts(pages)
+	if err != nil {
+		return nil, err
+	}
+
+	var networkIDs []string
+	for _, port := range ports {
+		if port.NetworkID != "" {
+			networkIDs = append(networkIDs, port.NetworkID)
+		}
+	}
 
-	err := pager.EachPage(func(page pagination.Page) (bool, error) {
-		s, err := servers.ExtractServers(page)
-		if err != nil {
-			return false, err
+	return networkIDs, nil
+}
+
+func getNodeNameByAddr(addr string, nodes []*v1.Node) (types.NodeName, bool) {
+	for _, node := range nodes {
+		for _, v := range node.Status.Addresses {
+			if v.Address == addr {
+				return types.NodeName(node.Name), true
+			}
 		}
-		for _, server := range s {
-			ok, err := handler(&server)
-			if !ok || err != nil {
-				return false, err
+	}
+
+	klog.V(4).Infof("Unable to resolve node name by %s IP", addr)
+	return types.NodeName(addr), false
+}
+
+func getAddrByNodeName(name types.NodeName, needIPv6 bool, nodes []*v1.Node) string {
+	for _, node := range nodes {
+		if node.Name == string(name) {
+			for _, v := range node.Status.Addresses {
+				if v.Type == v1.NodeInternalIP {
+					ip := net.ParseIP(v.Address)
+					if ip == nil {
+						continue
+					}
+					if needIPv6 && ip.To4() == nil {
+						return v.Address
+					}
+					return v.Address
```

Comment:
This if-branch is ineffective because the return value is the same regardless of whether the condition is true.

Reply:
Thanks, good catch!
Fixed in atomic routes PRs.